cloud/openstack/os_keystone_endpoint.py

```diff
@@ -0,0 +1,334 @@
+        enabled = "True" == module.params.get('enabled', True)
+        state = module.params.get('state', None)
+        admin_url = module.params.get('admin_url', None)
+        internal_url = module.params.get('internal_url', None)
+        public_url = module.params.get('public_url', None)
+
+        services_list = cloud.search_services(
+            filters={'name': service_name, 'type': service_type})
+        if len(services_list) != 1:
+            module.fail_json(
+               msg='Service %s not found or more than 1 service have this name'
+                    % service_name)
+        # Only service id is capable to identify unique entry.
+        service_id = services_list[0].id
+
+        current_endpoint = get_endpoint(service_id, interface)
+
+        if current_endpoint is None and state == 'absent':
+            module.exit_json(changed=False)
+
+        # Check if changes.
+        changed = False
+        if current_endpoint is None:
+            changed = True
+        else:
+            if url and (enabled is not None
+                        and current_endpoint.enabled != enabled):
+                changed = True
+            if hasattr(current_endpoint, 'region') \
+                    and current_endpoint.region != region or \
+                    not hasattr(current_endpoint, 'region') \
+                    and region is not None:
+                changed = True
+            if url:
+                if current_endpoint.url != url:
+                    changed = True
+            else:
+                if hasattr(current_endpoint, 'internalurl') \
+                        and current_endpoint.internalurl != internal_url \
+                        or not hasattr(current_endpoint, 'internalurl') \
+                        and internal_url is not None:
+                    changed = True
+                if hasattr(current_endpoint, 'adminurl') \
+                        and current_endpoint.adminurl != admin_url \
+                        or not hasattr(current_endpoint, 'adminurl') \
+                        and admin_url is not None:
+                    changed = True
+                if hasattr(current_endpoint, 'publicurl') \
+                        and current_endpoint.publicurl != public_url \
+                        or not hasattr(current_endpoint, 'publicurl') \
+                        and public_url is not None:
+                    changed = True
+
+        # Handle check mode.
+        if module.check_mode:
+            if changed and state == "absent":
+                changed = False
+            module.exit_json(changed=changed)
+
+        args = {}
+        for endpoint in ["public_url", "internal_url", "admin_url"]:
+            args[endpoint] = module.params.get(endpoint)
+
+        # Do changes.
+        if changed:
+            if state == "present":
+                if current_endpoint is None:
+                    cloud.create_endpoint(service_id,
```

Comment:
due to the way shade works, it is safe to pass interface, url, public_url, admin_url, and internal_url for each request, as long as they are set to None. That just means the logic for mutually_exclusive bits must be handled appropriately in the ansible module and then we can safely pass them all through.